Allow uploading large snapshots

sabeechen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This changes the endpoint that proxys commands to the supervisor by making it always stream request data to/from the requesting client and the supervisor, rather than reading it into memory.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
The goal for this change is to make the installation/onboarding process a little easier by allowing larger snapshots to be uploaded without having to find another way to get a large snapshot onto the filesystem.  Once I've made the relevant change in the frontend, snapshot upload will look something like this:
<img src="https://user-images.githubusercontent.com/21026774/127066516-b5f3eced-0ad9-4468-af53-ab63937dc478.gif" height="400px">
(I'm writing these frontend changes at the moment)


Some additional notes about this change:
* I've also added some extra test coverage to components/hassio/text_http.py to get it up to a 100% reported coverage.
* I've had a successful but very difficult time getting a full Supervisor + Core + Frontend stack running with local changes inside of my devcontainer.  If you have any guidance on how this is normally accomplished more easily I'd be very interested.  I plan to contribute more to these projects in the future.
* I couldn't determine why reading from the client in this proxy method was wrapped in an async_timeout.timeout() instead of just relying on aiohttp's normal read timeout handling, thought I did track it back to [this change](https://github.com/home-assistant/core/pull/6935/files#diff-b7d29db8f27322f022d9a53a9d853e41aabf104b735b3e4f284081df877d4f1cR182) originally.  Removing it was necessary to accomplish this change.
* The frontend still enforces a limit on the size of snapshot uploads client-side [here](https://github.com/home-assistant/frontend/blob/dev/hassio/src/components/hassio-upload-snapshot.ts#L50), so from users perspective this change will be a no-op until I've made changes there.
* Not all tests on a clean install passed without these changes, but I have verified that this change doesn't cause any additional test failures.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

sabeechen committed:
Merge pull request #2 from home-assistant/dev

Latest changes from home-assistant/core:dev

sabeechen committed:
Always stream proxy requests to the supervisor

pvizeli commented:
Yeah, it's similar to what we doing on ingress. I'm fine with the PR, just that with the timeout removal is a bit odd because now it takes 5minutes if the supervisor is not running until the user gets a response. Can we add an aiohttp timeout object with a limited connection timeout but have an default total timeout? In case the supervisor is not running, the user gets fast featback.

sabeechen committed:
Add connect timeout to proxied supervisor requests

sabeechen committed:
Merge branch 'dev' into stream-supervisor

sabeechen commented:
> Yeah, it's similar to what we doing on ingress. I'm fine with the PR, just that with the timeout removal is a bit odd because now it takes 5minutes if the supervisor is not running until the user gets a response. Can we add an aiohttp timeout object with a limited connection timeout but have an default total timeout? In case the supervisor is not running, the user gets fast featback.

That makes sense, I've added a 5 second connect time out on requests to the supervisor, let me know if you think it should be something different.  Noteably, aiohttp's request mocking doesn't track timeouts so I wasn't able to think of a good way to actually verify this change in tests.

There may be a misunderstanding.  The timeout I've removed is from [core's webserver reading data from the client](https://github.com/home-assistant/core/blob/dev/homeassistant/components/hassio/http.py#L100) (eg the browser), not on reading data from the supervisor.  This seemed like a bad idea to me, because the timeout would get reported back as 502 Bad Gateway and slow browsers (eg over bad WiFi) could have requests fail even if they're just sending data slowly.
  Looking through the file's commit history I couldn't figure out why it was added in the first place, so I wanted to call it out in case I'm breaking an undocumented assumption somewhere.

IIUC the way requests are proxied right now will *already* take 5 minutes if the supervisor is down, thought the connect timeout change I've made should change that.

pvizeli commented:
Let's make 10sec because a system can be busy during startup. After that we can merge it 👍

pvizeli commented on homeassistant/components/hassio/http.py:
```diff
@@ -151,11 +125,11 @@ def _init_header(request: web.Request) -> dict[str, str]:
     if NO_TIMEOUT.match(path):
-        return 0
-    return 300
+        return ClientTimeout(connect=5)
+    return ClientTimeout(connect=5, total=300)
```
```suggestion
        return ClientTimeout(connect=10)
    return ClientTimeout(connect=10, total=300)
```

sabeechen committed:
Change supervisor connect timeout to 10s

Co-authored-by: Pascal Vizeli <pascal.vizeli@syshack.ch>

sabeechen commented:
10 seconds sgtm, and thanks for the quick review!

ludeeus commented on homeassistant/components/hassio/http.py:
```diff
@@ -75,48 +72,28 @@ async def _handle(
-        client_timeout = 10
-        data = None
         headers = _init_header(request)
         if path in ("snapshots/new/upload", "backups/new/upload"):
             # We need to reuse the full content type that includes the boundary
             headers[
                 "Content-Type"
             ] = request._stored_content_type  # pylint: disable=protected-access
```
Is this override still needed with this approach?

sabeechen replied:
This PR didn't change anything about how the request looks to the supervisor (just how it sends over the bytes), so if it was needed before its still needed.
If I'm understanding correctly *why* this is needed, to include the file form data boundary parameter, I think it could come from request.raw_headers and doesn't need be specific to the /*/new/upload methods.

balloob commented:
We've reverting this PR because it broke ingress and prevented people from visiting the supervisor panel to upgrade/downgrade.

sabeechen commented:
Thats unfortunate, and a little disconcerting.  I wasn't able to find any communication problems when I synced my repos and ran a local HA+supervisor with the changes.
Could you provide any information that might point me toward making it surface in a development environment?  I'd like to make a new PR with fixes and some additional coverage to prevent whatever it is in the future.

ludeeus commented:
The response headers needs to be adjusted.
This diff makes it work (copied from the supervisor repo)
```
diff --git a/homeassistant/components/hassio/http.py b/homeassistant/components/hassio/http.py
index 73e5549be9..3fb5e3f80e 100644
--- a/homeassistant/components/hassio/http.py
+++ b/homeassistant/components/hassio/http.py
@@ -8,7 +8,12 @@ import re
 import aiohttp
 from aiohttp import web
 from aiohttp.client import ClientError, ClientTimeout
-from aiohttp.hdrs import CONTENT_TYPE
+from aiohttp.hdrs import (
+    CONTENT_ENCODING,
+    CONTENT_LENGTH,
+    CONTENT_TYPE,
+    TRANSFER_ENCODING,
+)
 from aiohttp.web_exceptions import HTTPBadGateway
 
 from homeassistant.components.http import KEY_AUTHENTICATED, HomeAssistantView
@@ -94,7 +99,9 @@ class HassIOView(HomeAssistantView):
             )
 
             # Stream the supervisor response back
-            response = web.StreamResponse(status=client.status, headers=client.headers)
+            response = web.StreamResponse(
+                status=client.status, headers=_response_header(client)
+            )
             response.content_type = client.content_type
 
             await response.prepare(request)
@@ -125,6 +132,23 @@ def _init_header(request: web.Request) -> dict[str, str]:
     return headers
 
 
+def _response_header(response: aiohttp.ClientResponse) -> dict[str, str]:
+    """Create response header."""
+    headers = {}
+
+    for name, value in response.headers.items():
+        if name in (
+            TRANSFER_ENCODING,
+            CONTENT_LENGTH,
+            CONTENT_TYPE,
+            CONTENT_ENCODING,
+        ):
+            continue
+        headers[name] = value
+
+    return headers
+
+
 def _get_timeout(path: str) -> ClientTimeout:
     """Return timeout for a URL path."""
     if NO_TIMEOUT.match(path):
```

sabeechen commented:
Thanks for the recommended fix, though I'm still unable to reproduce this problem in a way that could let me verify it actually fixes things.  I'm not sure what I'm trying to fix.
I've spent a lot of time trying to set up dev environment that lets me test local HA changes inside the supervisor (using a supervisor devcontainer), but it seems I'm still not making something the reproduces a "real enough" environment to catch issues I might introduce.
I'm at the point where I'm going to have to drop this PR for now for the sake of time and frustration.  I think this is a task better suited to someone who already knows how these components are developed together, or when I'm able to invest more time into sorting it out myself.

If someone would like to pick up this effort in the future, I've also written some much more substantial changes that adds the frontend changes necessary to make a progress bar for snapshot uploads, which is what I used to make the video shown in this PR.  If someone reading this in the future is interested in that, I think it would make an excellent starting point.  Almost an end point, actually.

I've kept those changes in a [draft PR here](https://github.com/sabeechen/frontend/pull/1) on my frontend fork, which I'll leave up.  Attribution is not important to me if someone else chooses to take this over.

ludeeus commented:
Have reimplemented this in https://github.com/home-assistant/core/pull/53909 with my suggestion included.
